### ISSUE TITLE
[WinHttpHandler OOB] Override APM (Begin/End) Stream APIs for WinHttpResponseStream

### ISSUE THREAD

**Issue by CIPop** [dotnet/corefx#11910]
On .Net Desktop, `StreamContent.SerializeToStreamAsync` is calling into `StreamToStreamCopy` which relies on Begin/End Stream APIs. This is different from CoreFx where StreamToStreamCopy is using Tasks.

Since on .Net Desktop the `System.IO.Stream` defaults are used, the code eventually calls synchronous Read/Write leading to very low performance when multiple HttpClients are used in parallel.

This is not an issue when using the System.Net.Http contract from CoreFX (AFAIK not possible in a .Net Desktop application).

(12 logical CPU box, using Parallel.For to schedule multiple HttpClients:)
![image](https://cloud.githubusercontent.com/assets/8635911/18686791/20bf7548-7f32-11e6-9e86-30b8d7ba4ac0.png)

**Comment by davidsh** [comment]
>This is not an issue when using the System.Net.Http contract from CoreFX (AFAIK not possible in a .Net Desktop application).

One can use System.Net.Http 4.1 contract from CoreFx on .NET Framework (Desktop).  Simply reference the System.Net.Http 4.1 contract in project.json and you will end up getting the OOB'd System.Net.Http binary.

**Comment by davidsh** [comment]
cc: @stephentoub

**Comment by CIPop** [comment]
Discussed with @davidsh offline. This bug affects .Net Framework performance only (not Core) but requires changes in the CoreFX repo which builds WinHttpHandler.

The repro involves importing _only_ the System.Net.Http.WinHttpHandler package and evaluating performance using it in conjunction with System.Net.Http.HttpClient from .Net Framework. (Can't mark up for grabs as CoreFX/NetFX OSS perf testing is still WIP.)

We consider this a priority for .Net Framework as it is the only supported way to use HTTP2.0 but not a priority for Standard 2.0.

**Comment by stephentoub** [comment]
Is the issue here that WinHttpResponseStream needs to override Begin/EndRead?  If so, it does:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs#L202
https://github.com/dotnet/corefx/pull/16134

**Comment by CIPop** [comment]
Cool! Then this should be fixed! (Will validate once we're done with our perf story.)
